### PR TITLE
[RFC] Output data type

### PR DESCRIPTION
**Pull request by saqadri**
[RFC] Output data type

Add some structure to the outputs that allow the frontend to know how to render them.

Still need to update the Python types, and modelparsers to return data in this type.

### CODE REVIEW

File: schema/aiconfig.schema.json [hunk 1]
```diff
@@ -174,27 +191,6 @@
         }
       ]
     },
-    "Attachment": {
```
Comment (rholinshead) [inline schema/aiconfig.schema.json:177]: hmm, why the change to make this inline instead of its own definition?

File: schema/aiconfig.schema.json [hunk 2]
```diff
@@ -157,15 +157,32 @@
...
+                    "additionalProperties": {}
+                  }
+                },
+                "required": [
+                  "data"
+                ]
               }
+            },
+            "data": {
+              "description": "Input to the model. This can represent a single input, or multiple inputs.\nThe structure of the data object is up to the ModelParser. For example,\na multi-modal ModelParser can choose to key the data by MIME type."
```
Comment (rholinshead) [inline schema/aiconfig.schema.json:185]: Why the change to the descriptions here? I think it's pretty confusing to reference mime-types here instead of having an opinionated stance on using attachments for different mime type data -- reading these descriptions I wouldn't know which to pick and it wouldn't seem to matter?

File: typescript/types.ts [hunk 3]
```diff
@@ -192,8 +192,12 @@ export type ExecuteResult = {
   /**
    * The result of executing the prompt.
    */
-  data: JSONValue;
-
+  data:
+    | JSONValue
+    | {
```
Comment (rholinshead) [inline typescript/types.ts:197]: The thing I'm not sure of here is if we should have this data associated with an Attachment so we can know its mimetype. Even if we have file_uri, we still need to know what to render it as (image, video, audio, etc.)

File: typescript/types.ts [hunk 4]
```diff
@@ -192,8 +192,12 @@ export type ExecuteResult = {
...
+        kind: "string" | "file_uri" | "base64";
+        value: string;
+      };
```
Comment (rholinshead) [inline typescript/types.ts:200]: Ah, right, I guess we just rely on the mime_type below to know the represented type

File: typescript/types.ts [hunk 5]
```diff
@@ -192,8 +192,12 @@ export type ExecuteResult = {
   /**
    * The result of executing the prompt.
    */
-  data: JSONValue;
-
+  data:
+    | JSONValue
```
Comment (rholinshead) [inline typescript/types.ts:196]: Long-term, do we want to support arbitrary JSONValue here or just let them dump things into metadata so we can always have a kind/value?

File: schema/aiconfig.schema.json (hunk 1 above)
Comment (saqadri) [reply schema/aiconfig.schema.json:177]: I'm not sure -- this was autogenerated schema based on the types.ts

File: typescript/types.ts (hunk 5 above)
Comment (saqadri) [reply typescript/types.ts:196]: We can decide that. My 2c are it's still good to support arbitrary JSONValue here, but we can revisit that. It will require a breaking change to the schema too, so we can consider this for v2 schema

File: schema/aiconfig.schema.json (hunk 1 above)
Comment (saqadri) [reply schema/aiconfig.schema.json:177]: It looks like we may have manually updated the schema last time, and this is being updated with the `genSchema` script.

File: schema/aiconfig.schema.json (hunk 2 above)
Comment (saqadri) [reply schema/aiconfig.schema.json:185]: I think we didn't update the schema using the script for a while. I've updated the description to remove references to MIME types

File: python/src/aiconfig/schema.py [hunk 6]
```diff
@@ -11,20 +11,37 @@
...
     mime_type: Optional[str] = None
     # Output metadata
     metadata: Dict[str, Any]
 
 
 class Error(BaseModel):
+    """
+    Error represents an error that occurred while executing a prompt.
+    """
+
```
Comment (saqadri) [inline python/src/aiconfig/schema.py:44]: These are the only changes -- the rest are autoformatting changes

File: schema/aiconfig.schema.json (hunk 1 above)
Comment (rholinshead) [reply schema/aiconfig.schema.json:177]: Ah, that was definitely me. I didn't realize there's a gen script. We should update the README to reference that and the process -- just make changes to types.ts and run `yarn gen-schema`?

Comment (rholinshead) [reply schema/aiconfig.schema.json:177]: I guess the gen script removes the definition/ref for some reason. Not a big deal

**Review by rholinshead — APPROVED** [review]

**Comment by saqadri** [comment]
@rossdanlm I'm going to ship this -- can you take up the model parser updates to satisfy these types. Can start with a couple of them (maybe Dalle and GPT) to see what it's like. cc @rholinshead
